packages/hardhat-zksync-solc/src/compile/docker.ts

```diff
@@ -0,0 +1,117 @@
+import {
+    DockerBadGatewayError,
+    DockerHubConnectionError,
+    DockerNotRunningError,
+    DockerServerError,
+    HardhatDocker,
+    Image,
+    ImageDoesntExistError,
+    ProcessResult,
+} from '@nomiclabs/hardhat-docker';
+import { ProjectPathsConfig, CompilerInput } from 'hardhat/types';
+import path from 'path';
+import { ZkSolcConfig } from '../types';
+
+import { pluginError } from '../utils';
+
+// Notice: contents of this file were mostly copy-pasted from the
+// official Hardhat Vyper plugin: https://github.com/nomiclabs/hardhat/tree/master/packages/hardhat-vyper
+
+export function dockerImage(imageName?: string): Image {
+    if (!imageName) {
+        throw pluginError('Docker source was chosen but no image was specified');
+    }
+
+    return {
+        repository: imageName,
+        tag: 'latest',
+    };
+}
+
+export async function validateDockerIsInstalled() {
+    if (!(await HardhatDocker.isInstalled())) {
+        throw pluginError(
+            `Docker Desktop is not installed.
+Please install it by following the instructions on https://www.docker.com/get-started`
+        );
+    }
+}
+
+export async function createDocker(): Promise<HardhatDocker> {
+    return await handleCommonErrors(HardhatDocker.create());
+}
+
+export async function pullImageIfNecessary(docker: HardhatDocker, image: Image) {
+    await handleCommonErrors(pullImageIfNecessaryInner(docker, image));
+}
+
+async function pullImageIfNecessaryInner(docker: HardhatDocker, image: Image) {
+    if (!(await docker.hasPulledImage(image))) {
+        console.log(`Pulling Docker image ${HardhatDocker.imageToRepoTag(image)}...`);
+
+        await docker.pullImage(image);
+
+        console.log(`Image pulled`);
+    } else {
+        await checkForImageUpdates(docker, image);
+    }
+}
+
+async function checkForImageUpdates(docker: HardhatDocker, image: Image) {
+    if (!(await docker.isImageUpToDate(image))) {
+        console.log(`Updating Docker image ${HardhatDocker.imageToRepoTag(image)}...`);
+
+        await docker.pullImage(image);
+
+        console.log(`Image updated`);
+    }
+}
+
+export async function compileWithDocker(
+    input: CompilerInput,
+    docker: HardhatDocker,
+    dockerImage: Image,
+    config: ZkSolcConfig
+): Promise<any> {
+    const zksolcCommand = ['zksolc', '--standard-json'];
+    if (config?.settings?.optimizer?.enabled) {
+        zksolcCommand.push('--optimize');
+    }
+    // TODO: pass the input into docker
+
+    const result = handleCommonErrors(docker.runContainer(dockerImage, zksolcCommand));
```

Comment:
Does it mean that docker does not work yet? I don't think we can merge it without docker support.

Reply:
Yup, that's the last thing left to do. Working on it

Reply:
From when I was looking into this, supporting docker requires breaking up all of the clean `TASK`s that hardhat uses and the code will look a lot more like my PR. Because you have to initialize the docker compiler early on and pass that through the tasks, otherwise if you run `compileWithDocker` within the loop of  `TASK_COMPILE_SOLIDITY_RUN_SOLC` docker will start rate limiting and blocking due to all of the image update checks.

Tbh as an end user I'd prefer the docker support to be dropped and instead prioritize getting a cross platform binary working. But obviously I don't have all of the context, just my 2 cents

Reply:
> I'd prefer the docker support to be dropped and instead prioritize getting a cross platform binary working

That's planned, but for now, while the compiler is in an active development phase, it's much easier to distribute updates currently. Once we have a stable release, native binaries will surely be a priority.